worker/sshserver/listener.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshserver
+
+import (
+	"net"
+	"sync"
+)
+
+// acceptOnceListener is required to prevent a race condition
+// that can occur in tests.
+//
+// The SSH server tracks the listeners in use
+// but if the server's close() method executes
+// before we reach a safe point in the Serve() method
+// then the server's map of listeners will be empty.
+// A safe point to indicate the server is ready is
+// right before we start accepting connections.
+// Accept() will return with error if the underlying
+// listener is already closed.
+//
+// As such, we ensure accept has been called at least once
+// before allowing a close to take effect. The corresponding
+// piece to this is to receive from the closeAllowed channel
+// within your cleanup routine.
+type acceptOnceListener struct {
```

Comment:
name does not really reflect its function.. its purpose is to block the close call until we've reached the safe point in the ssh server, where accept is called.

Reply:
True... I'm bad at names but it was called sync originally which isn't right either